common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,504 @@
+// The MIT License
+//
+// Copyright (c) 2025 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testlogger
+
+import (
+	"container/list"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zaptest"
+)
+
+type Level string
+
+const (
+	Debug Level = "debug"
+	Info  Level = "info"
+	Warn  Level = "warn"
+	// Panics, DPanics, and Fatal logs are considered errors
+	Error Level = "error"
+)
+
+// This is a subset of the testing.T interface that we use in this package that is
+// also shared with *rapid.T.
+type TestingT interface {
+	zaptest.TestingT
+	Helper()
+	Name() string
+	Logf(format string, args ...any)
+	Log(args ...any)
+	Fatalf(format string, args ...any)
+	Fatal(args ...any)
+}
+
+// CleanupCapable is an interface that allows a test to register cleanup functions.
+// This is /not/ implemented by *rapid.T but is by *testing.T
+type CleanupCapableT interface {
+	TestingT
+	Cleanup(func())
+}
+
+func (l Level) String() string {
+	return string(l)
+}
+
+// Expectations represent errors we expect to happen in tests.
+// Their only purpose is to allow un-expecting (hah!) an error we've
+// marked as expected
+type Expectation struct {
+	e          *list.Element
+	testLogger *TestLogger
+	lvl        Level
+	unexpected bool
```

Comment:
Change from original: added this to denote expected und "unexpected" expectations.

PS: I considered renaming it to `Rule` but stuck with the current name.